README
The CTk Toolkit

by Martin Andrews
andrewm@ccfadm.eeg.ccf.org

@(#) $Id: ctk.shar,v 1.50 1996/01/15 14:47:16 andrewm Exp andrewm $

1. Introduction
----------------

This directory contains the sources for CTk, a curses port of John
Ousterhout's Tk toolkit for X11.  The information here corresponds to
CTk 4.0.  CTk 4.0 is based on Tk 4.0p3. Like Tk 4.0p3, it is designed
to work with Tcl 7.4p3 and may not work with other releases of Tcl.

Using CTk, applications with a modern GUI-ish interface can be created
for character terminals.  These same applications, without modification,
can provide a real GUI interface by using Tk.  Thus, sites with an
embedded base of character terminals (and a small capital budget) can
smoothly migrate to GUI applications.

2. Documentation
----------------

There is a man page for cwish (cwish(1)) explaining execution options
for the CTk shell.  For script writing, the Tk Documentation, along with the
list of differences in section 5 of this document should be enough to get
started.

3. Compiling and installing CTk
-------------------------------

CTk requires TCL and a System V curses package (CTk will build
with BSD curses but has at least one display glitch and many
missing features - see porting.notes for more information).
Unlike Tk, you do not need the source for Tcl to build CTk.
Only the Tcl 7.4 library and include file (tcl.h) are necessary.
If your system does not have a System V compatible curses,
there is a free version, ncurses, available in GNU archives
(for instance ftp://prep.ai.mit.edu/pub/gnu).

CTk builds cleanly on the three systems I have easy access to:

    486 PC running Linux 1.1.54 using gcc 2.58 and ncurses 1.8.5

    HP 9000/835 running HP-UX 9.04 using optional ANSI C compiler and
    ncurses 1.8.6 or using K&R C compiler and hp curses (old system V,
    not very pretty)

    Sun 4c running Sun OS 4.13 using System V compiler (/usr/5bin/cc)
    and System V curses

Perform the following steps to compile and install CTk:

    (a) Type "./configure" in this directory.  This runs a configuration
	script created by GNU autoconf, which configures CTk for your
	system and creates a Makefile.  If you are using ncurses, and
	the library is named "libncurses" instead of "libcurses" give
	the "-with-ncurses" option to configure.  Also, if Tcl or curses
	are not installed in the standard search path, you can use
	"-with-libdirs=" and "-with-incdirs=" to add directories to
	the library and header search path, respectively (use the
	directives once, the argument can be a list of space separated
	directories).  For more details on using configure, check out
	the autoconf documentation (not included here).

    (b) Type "make".  This will create a library archive called "libctk.a"
	and an interpreter application called "cwish" that allows you to type
	Tcl commands interactively or execute script files.

    (c) If the make fails then you'll have to personalize the Makefile
        for your site or possibly modify the distribution in other ways.
	First check the file "porting.notes" to see if there are hints
	for compiling on your system.  If you need to modify Makefile,
	there are comments at the beginning of it that describe the things
	you might want to change and how to change them.
	
    (d) Type "make install" to install CTk's binaries and script files in
        standard places.  In the default configuration, information will
	be installed in /usr/local so you'll need write permission on
	this directory.  If you'd like to use a default installation
	directory, you can change the "exec_prefix" and "prefix" definitions
	in the Makefile.

    (e) At this point you can play with Tcl by invoking the "cwish"
	program and typing Tcl commands.  However, if you haven't installed
	CTk then you'll first need to set your CTK_LIBRARY environment
	variable to hold the full path name of the "library" subdirectory.

I am interested in receiving information on changes required to
build CTk on your platform.

4. Test suite
-------------

Next release.  (Really.)

5. Getting started
------------------

It is now easy to run cwish interactively.  If you have not
installed cwish set the CTK_LIBRARY environment variable to
the path of the library directory, then try:

    cwish

This will start cwish, and then pop-up a command dialog window where
you can enter TCL commands.  You can bring up this dialog at any time
by pressing <Control-C>.

You can also try out a simple demo of the cwish widgets with:

    cwish library/demos/stuff

There are other demos in the library/demos directory including a
crude port of the Tk widget demo (warning it is easy for the focus
to get lost - if so press Ctrl+C to get the command dialog and
type "exit" there).

Key bindings are very nearly the same as for Tk (which is very nearly
the same as Motif).  This may be unfamiliar to people used to common
Unix curses applications (elm, lynx, ..).  I decided to stick with the
Tk binding for two reasons:

1)  Less confusion for users that switch between Tk and CTk
    versions of an application.

2)  Easier to incorporate updates from Tk.

Here is a quick summary of the key bindings for those unfamiliar
with Motif (or Microsoft Windows):

    <Space> activates/selects a widget
    <Tab> moves focus to next the widget
    <Shift-Tab> moves focus to the prior widget
    <Return> activates the default button in a dialog
    Arrow keys move the cursor within a widget
    <F10> moves focus to the first menu
    <Escape> closes a menu without making a selection

Your terminal or terminfo entry may not support all the keys (like
<Shift-Tab>).  See README.TERM for tips on configuring terminfo
entries for CTk.

6. Summary of differences between CTk 4.0b1 and Tk 4.0
--------------------------------------------------------

The following commands are not available in CTk:

    canvas
    clipboard
    message
    image
    scale
    selection
    send

The wm command is severely crippled.

None of the widgets support the scan method.

The text widget does not support tag bindings: "text tag bind".

The text widget does not support embedded windows "text window".

The -tearoff option for menu widgets can create a tearoff entry,
but the entry doesn't work (and I don't know if there is any point
in making it work).

Many configuration options (like -background and -foreground) cannot
be modified.  Attempts to set the unsupported options will silently
be ignored.  Querying the options with cget will return a fixed
?reasonable? value.  The unsupported options are:

    -activebackground
    -activeborderwidth
    -activerelief
    -background
    -bg
    -bitmap
    -borderwidth (supported by widgets, not by text tags)
    -colormap
    -cursor
    -disabledforeground
    -exportselection
    -fg
    -font
    -foreground
    -highlightcolor
    -highlightbackground
    -highlightthickness
    -image
    -insertbackground
    -insertborderwidth
    -insertofftime
    -insertontime
    -insertwidth
    -indicatoron
    -jump
    -relief
    -repeatdelay
    -repeatinterval
    -screen
    -selectbackground
    -selectborderwidth
    -selectcolor
    -selectforeground
    -selectimage
    -setgrid
    -tearoff
    -troughcolor
    -visual

And a lot more I forgot to mention.

7. Support and bug fixes
------------------------

Send bug reports and suggestions for improvements to:

    Martin Andrews
    andrewm@ccfadm.eeg.ccf.org

When reporting bugs, please provide a short cwish script that I can
use to reproduce the bug.  Make sure that the script runs with a
bare-bones cwish and doesn't depend on any extensions.  Also, please
include three additional pieces of information with the script:

    (a) how do I use the script to make the problem happen (e.g.
	what keys do you press, in what order)?
    (b) what happens when you do these things (presumably this is
        undesirable)?
    (c) what did you expect to happen instead?

For general problems with using Tk or Tcl, try posting to the
comp.lang.tcl Usenet newsgroup.

8. Release organization
---------------------------

Each CTk release is identified by two numbers separated by a dot, e.g.
3.2 or 3.3.  These numbers match the release number of the corresponding
Tk release.  Suffixes for alpha, beta, and patch releases (aX, bX, and pX
respectively) are numbered independent of Tk.


changes
Changes made since ctk8.0 was initially released are summarized below:

Changed the 'configure' command to recognize tcl8.0
Changed the Makefile.in to build cwish$(VERSION) instead of cwish

This file summarizes all changes made to CTk since version 4.0a1 was
released on March 12, 1995.  Changes that aren't backward compatible
are marked specially.
 
@(#) $Id: ctk.shar,v 1.50 1996/01/15 14:47:16 andrewm Exp andrewm $

$Log: ctk.shar,v $
# Revision 1.50  1996/01/15  14:47:16  andrewm
# More portability glitches (needed to reset interpreter result
# in TkEntryInsert routines).
#
# Revision 1.49  1996/01/13  18:31:51  andrewm
# Dumb typos in tkCmds.c - found when build under SunOS.
#
# Revision 1.48  1995/12/21  19:31:05  andrewm
# Added "ctk redraw" command.
#
# Revision 1.47  1995/12/20  22:22:00  andrewm
# (bug fix) made ctkDisplay.c 8-bit clean (wasn't casting char
# to unsigned char) - thanks Jan Vlcek.
#
# Revision 1.46  1995/12/20  21:13:04  andrewm
# Incorporate changes from Tk4.0 patch 3 (tk4.0p3).
#
# Revision 1.45  1995/12/20  16:28:09  andrewm
# Incorporate changes from tk4.0 patch 2.
#
# Revision 1.44  1995/12/04  19:57:27  andrewm
# (bug fix) tkListbox.c - display cursor even when active row is not
# visible (put it in the top left corner).
#
# (feature change) tkListbox.c - no longer display "*" next to selected
# items - rely on display attribute instead.
#
# (feature change) listbox.tcl - Left and Right now scroll by pages
# instead of characters.
#
# Revision 1.43  1995/12/04  19:19:52  andrewm
# bug fix - forgot to change var name when pasting code into tk_popup.
#
# Revision 1.42  1995/11/19  17:07:16  andrewm
# Incorporate changes from Tk4.0 patch 1.
#
# Revision 1.41  1995/11/17  15:35:49  andrewm
# Drop use of A_INVIS in ctkDisplay.c - those styles should never be
# drawn anyway - and most terms don't support it (and it screws up
# out terminal session manager).
#
# Revision 1.40  1995/11/06  13:31:19  andrewm
# Change TK_PATCH_LEVEL to CTK_PATCH_LEVEL and tcl variable tk_patchLevel
# with ctk_patchLevel.  Only version numbers track between Tk and CTk -
# and this make for an easy way to tell when you are using CTk.
#
# Add Tk_RestackWindow() and pull stacking arguments out of Tk_Map() -
# now raise and lower commands use Tk_RestackWindow instead of Ctk_Map
# (because they should not map unmapped windows).
#
# The CTk menu now underlines the X in Exit (not the E).
#
# Revision 1.39  1995/08/31  19:48:54  andrewm
# Make sure that all uses of ckalloc/ckfree are with (char *) pointers
# (cast when necessary).
#
# Install as version 4.0 - don't need a different library for each patch.
#
# Revision 1.38  1995/08/25  20:38:47  andrewm
# bug fix - finally track down glitch that caused some slaves to not
# be mapped - backwards logic in tkGeometry.c.
#
# aesthetics - tk_dialog now uses a simpler appearance for the default
# button (add [] to label instead of an enclosing frame).
#
# Revision 1.37  1995/08/25  18:41:19  andrewm
# Change tkFindMenu to that it only finds menubuttons that do not
# have an indicator (otherwise the F10 key will fire menus of
# options buttons).
#
# Add exit to Ctk menu.
#
# Revision 1.36  1995/08/24  17:50:35  andrewm
# Add compatibilty info to tk.h to simplify porting of Tk extensions
# to ctk (BLT table was easy!).
#
# Revision 1.35  1995/08/22  19:48:01  andrewm
# Fix bug when destroying all windows from within a command from
# an event binding (could core dump).
#
# Revision 1.34  1995/08/22  17:31:37  andrewm
# Add ability for multi-column (wrapped) menus.
#
# Change menu button indicator to "^" (was "=").
#
# Revision 1.33  1995/08/21  21:39:57  andrewm
# More logical key fixes - no longer bind directly to "space" - instead
# bind to Select, and remap space to Select.
#
# Make Return a do-what-I-want key instead of the dialog closure key
# (which is now Execute).
#
# Curses Enter/Send key now maps to KP_Enter, not Return.
#
# Revision 1.32  1995/08/21  20:47:50  andrewm
# Remove root window from name table after creating "." window -
# otherwise user can reference the root, which can cause core dumps.
#
# Change bindings to use logical key names only - then central
# ctk.tcl binds other keys to generate logical key events.
#
# Add ctk_event command (ala testevent).
#
# Revision 1.31  1995/08/21  14:52:49  andrewm
# Fix foolish change to text/entry insertion - back to Insert procefdure
# returning a boolean.
#
# Revision 1.30  1995/08/21  13:30:49  andrewm
# Pull out the menu traversal by matching first label character - decide
# it better belongs in a library.
#
# Revision 1.29  1995/08/18  21:58:02  andrewm
# Modify bindings within menu to activate entries by pressing the first
# letter of their label.
#
# Revision 1.28  1995/08/18  19:56:34  andrewm
# Rewite the tkEntryInsert and tkEntrySeeInsert procedures in C.
# Not sure if the payoff is worth it this time.
#
# Revision 1.27  1995/08/18  14:15:35  andrewm
# Rewrite the tk_focusNext and tk_focusPrev procedures in C, because
# their overhead was annoying on loaded server machines - particularly
# if you overshoot a widget and have not Shift-Tab on your terminal!

--- ctk4.0b1 Released ---

# Revision 1.26  1995/07/25  14:07:28  andrewm
# (new feature) Add some of the Tk demos - begin fixing them for CTk.
#
# Clean up documentation for CTk 4.0b1 release.
#
# Revision 1.25  1995/07/21  21:19:00  andrewm
# (bug fix) fixes for non-ANSI compiler (SunOs).
#
# Revision 1.24  1995/07/20  21:26:44  andrewm
# (new feature) add -with-libdirs and -with-incdirs directives for
# configure.
#
# (new feature) unsupported Tk commands are now passed to a "ctk_unsupported"
# command - if this command is not defined then the command is ignored.
#
# Revision 1.23  1995/07/20  18:05:00  andrewm
# (bug fix) tkListbox.c - don't let cursor scroll horizontally with contents
# of listbox.
#
# Revision 1.22  1995/07/20  16:20:50  andrewm
# Fill in cwish man page.
# Include configuration changes from Tk4.0
#

tkInt.h
/*
 * tkInt.h (CTk) --
 *
 *	Declarations for things used internally by the Tk
 *	procedures but not exported outside the module.
 *
 * Copyright (c) 1990-1994 The Regents of the University of California.
 * Copyright (c) 1994 Sun Microsystems, Inc.
 * Copyright (c) 1994-1995 Cleveland Clinic Foundation
 *
 * See the file "license.terms" for information on usage and redistribution
 * of this file, and for a DISCLAIMER OF ALL WARRANTIES.
 *
 * @(#) $Header: /usrs/andrewm/work/RCS/ctk.shar,v 1.50 1996/01/15 14:47:16 andrewm Exp andrewm $
 */

#ifndef _TKINT
#define _TKINT

#ifndef _TK
#include "tk.h"
#endif
#include <tcl.h>

/*
 * One of the following structures is maintained for each display
 * containing a window managed by Tk:
 */

struct TkDisplay {
    /*
     * Maintained by ctkDisplay.c
     */
    char *name;			/* Name of display device. Malloc-ed. */
    char *type;			/* Device type. Malloc-ed. */
    ClientData display;		/* Curses's info about display. */
    Tcl_Channel chan;		/* Input channel for the device */
    int fd;			/* Input file descriptor for device. */
    FILE *inPtr;		/* Input file pointer for device. */
    TkWindow *cursorPtr;	/* Window to display cursor in. */
    int cursorX, cursorY;	/* Position in `cursWinPtr' to display
				 * cursor. */

    /*
     * Maintained by tkWindow.c
     */
    int numWindows;		/* Windows currently existing in display
    				 * (including root). */
    TkWindow *rootPtr;		/* Root window of display. */
    TkWindow *focusPtr;		/* Window that has the keyboard focus. */
    struct TkDisplay *nextPtr;	/* Next in list of all displays. */
};

/*
 * One of the following structures exists for each event handler
 * created by calling Tk_CreateEventHandler.  This information
 * is used by tkEvent.c only.
 */

struct TkEventHandler {
    unsigned long mask;		/* Events for which to invoke
				 * proc. */
    Tk_EventProc *proc;		/* Procedure to invoke when an event
				 * in mask occurs. */
    ClientData clientData;	/* Argument to pass to proc. */
    struct TkEventHandler *nextPtr;
				/* Next in list of handlers
				 * associated with window (NULL means
				 * end of list). */
};

/*
 * Tk keeps one of the following data structures for each main
 * window (created by a call to Tk_CreateMainWindow).  It stores
 * information that is shared by all of the windows associated
 * with a particular main window.
 */

struct TkMainInfo {
    int refCount;		/* Number of windows whose "mainPtr" fields
				 * point here.  When this becomes zero, can
				 * free up the structure (the reference
				 * count is zero because windows can get
				 * deleted in almost any order;  the main
				 * window isn't necessarily the last one
				 * deleted). */
    struct TkWindow *winPtr;	/* Pointer to main window. */
    Tcl_Interp *interp;		/* Interpreter associated with application. */
    Tcl_HashTable nameTable;	/* Hash table mapping path names to TkWindow
				 * structs for all windows related to this
				 * main window.  Managed by tkWindow.c. */
    Tk_BindingTable bindingTable;
				/* Used in conjunction with "bind" command
				 * to bind events to Tcl commands. */
    TkDisplay *curDispPtr;	/* Display for last binding command invoked
				 * in this application;  used only  by
				 * tkBind.c. */
    int bindingDepth;		/* Number of active instances of Tk_BindEvent
				 * in this application.  Used only by
				 * tkBind.c. */
    struct ElArray *optionRootPtr;
				/* Top level of option hierarchy for this
				 * main window.  NULL means uninitialized.
				 * Managed by tkOption.c. */
    struct TkMainInfo *nextPtr;	/* Next in list of all main windows managed by
				 * this process. */
};

/*
 * Pointer to first entry in list of all displays currently known.
 */

extern TkDisplay *tkDisplayList;

/*
 * Flags passed to TkMeasureChars:
 */

#define TK_WHOLE_WORDS		 1
#define TK_AT_LEAST_ONE		 2
#define TK_PARTIAL_OK		 4
#define TK_NEWLINES_NOT_SPECIAL	 8
#define TK_IGNORE_TABS		16

/*
 * Location of library directory containing Tk scripts.  This value
 * is put in the $tkLibrary variable for each application.
 */

#ifndef CTK_LIBRARY
#define CTK_LIBRARY "/usr/local/lib/ctk"
#endif

/*
 * Special flag to pass to Tk_CreateFileHandler to indicate that
 * the file descriptor is actually for a display, not a file, and
 * should be treated specially.  Make sure that this value doesn't
 * conflict with TK_READABLE, TK_WRITABLE, or TK_EXCEPTION from tk.h.
 */

#define TK_IS_DISPLAY	32

/*
 * The macro below is used to modify a "char" value (e.g. by casting
 * it to an unsigned character) so that it can be used safely with
 * macros such as isspace.
 */

#define UCHAR(c) ((unsigned char) (c))

/*
 * Miscellaneous variables shared among Tk modules but not exported
 * to the outside world:
 */

extern Tk_Uid			tkActiveUid;
extern void			(*tkDelayedEventProc) _ANSI_ARGS_((void));
extern Tk_Uid			tkDisabledUid;
extern TkMainInfo		*tkMainWindowList;
extern Tk_Uid			tkNormalUid;

/*
 * Internal procedures shared among Tk modules but not exported
 * to the outside world:
 */

extern void		TkBindEventProc _ANSI_ARGS_((TkWindow *winPtr,
			    XEvent *eventPtr));
extern void		TkComputeTextGeometry _ANSI_ARGS_((
			    char *string,
			    int numChars, int wrapLength, int *widthPtr,
			    int *heightPtr));
extern int		TkCopyAndGlobalEval _ANSI_ARGS_((Tcl_Interp *interp,
			    char *script));
extern Time		TkCurrentTime _ANSI_ARGS_((void));
extern int		TkDeadAppCmd _ANSI_ARGS_((ClientData clientData,
			    Tcl_Interp *interp, int argc, char **argv));
extern void		TkDisplayChars _ANSI_ARGS_((TkWindow *winPtr,
			    Ctk_Style style, char *string,
			    int numChars, int x, int y, int tabOrigin,
			    int flags));
extern void		TkDisplayText _ANSI_ARGS_((TkWindow *winPtr,
			    Ctk_Style style,
			    char *string, int numChars, int x, int y,
			    int length, Tk_Justify justify, int underline));
extern void		TkEventCleanupProc _ANSI_ARGS_((
			    ClientData clientData, Tcl_Interp *interp));
extern void		TkEventDeadWindow _ANSI_ARGS_((TkWindow *winPtr));
extern void		TkFocusDeadWindow _ANSI_ARGS_((TkWindow *winPtr));
extern int		TkFocusFilterEvent _ANSI_ARGS_((TkWindow *winPtr,
			    XEvent *eventPtr));
extern void		TkFreeBindingTags _ANSI_ARGS_((TkWindow *winPtr));
extern TkWindow *	TkGetFocus _ANSI_ARGS_((TkWindow *winPtr));
extern int		TkGetInterpNames _ANSI_ARGS_((Tcl_Interp *interp,
			    Tk_Window tkwin));
extern char *		TkInitFrame _ANSI_ARGS_((Tcl_Interp *interp,
			    Tk_Window tkwin, int toplevel, int argc,
			    char *argv[]));
extern int		TkMeasureChars _ANSI_ARGS_((
			    char *source, int maxChars, int startX, int maxX,
			    int tabOrigin, int flags, int *nextXPtr));
extern void		TkOptionClassChanged _ANSI_ARGS_((TkWindow *winPtr));
extern void		TkOptionDeadWindow _ANSI_ARGS_((TkWindow *winPtr));
extern void		TkQueueEvent _ANSI_ARGS_((TkDisplay *dispPtr,
			    XEvent *eventPtr));

extern void		TkDeleteMain _ANSI_ARGS_((TkMainInfo *mainPtr));

#define CtkIsDisplayed(tkwin)	(((tkwin)->flags)& CTK_DISPLAYED)

typedef void (CtkSpanProc) _ANSI_ARGS_((int left, int right, int y,
	ClientData data));

#define CtkSpanIsEmpty(left, right) ((left) >= (right))
#define CtkCopyRect(dr,sr) (memcpy((dr), (sr), sizeof(Ctk_Rect)))
#define CtkMoveRect(rect,x,y) \
	((rect)->left += (x), (rect)->top += (y), \
	(rect)->right += (x), (rect)->bottom += (y))
#define CtkSetRect(rect,l,t,r,b) \
	((rect)->left = (l), (rect)->top = (t), \
	(rect)->right = (r), (rect)->bottom = (b))


EXTERN int		CtkDisplayInit _ANSI_ARGS_((Tcl_Interp *interp,
			    TkDisplay *dispPtr, char *displayName));
EXTERN void		CtkDisplayEnd _ANSI_ARGS_((TkDisplay *dispPtr));
EXTERN void		CtkDisplayBell _ANSI_ARGS_((TkDisplay *dispPtr));

EXTERN void         CtkSetFocus _ANSI_ARGS_((TkWindow *winPtr));
EXTERN void	    Ctk_Forget _ANSI_ARGS_((Tk_Window tkwin));

EXTERN void	    CtkFillRegion _ANSI_ARGS_((TkDisplay *dispPtr,
			CtkRegion *rgn_ptr, Ctk_Style style, int ch));

EXTERN void	    CtkIntersectSpans _ANSI_ARGS_((int *left_ptr,
			int *right_ptr, int left2, int right2));
EXTERN void	    CtkIntersectRects _ANSI_ARGS_((Ctk_Rect *r1_ptr,
			CONST Ctk_Rect *r2_ptr));
EXTERN CtkRegion * CtkCreateRegion _ANSI_ARGS_((Ctk_Rect *rect));
EXTERN void	    CtkDestroyRegion _ANSI_ARGS_((CtkRegion *rgn));
EXTERN void	    CtkForEachIntersectingSpan _ANSI_ARGS_((
			CtkSpanProc *func,
			ClientData func_data, int left, int right, int y,
			CtkRegion *rgn));
EXTERN void	    CtkForEachSpan _ANSI_ARGS_((CtkSpanProc *func,
			ClientData func_data, CtkRegion *rgn));
EXTERN CtkRegion * CtkRegionMinusRect _ANSI_ARGS_((CtkRegion *rgn_id,
			Ctk_Rect *rect, int want_inter));
EXTERN void	    CtkUnionRegions _ANSI_ARGS_((CtkRegion *rgn1,
			CtkRegion *rgn2));
EXTERN void	    CtkRegionGetRect _ANSI_ARGS_((CtkRegion *rgn,
			Ctk_Rect *rect_ptr));
EXTERN int	    CtkPointInRegion _ANSI_ARGS_((int x, int y,
			CtkRegion *rgn));

#endif  /* _TKINT */
